File: Python/Linked_List/143.重排链表.py
# Definition for single-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverse(self, node):
        if node.next == None:
            return node
        last = self.reverse(node.next)
        node.next.next = node
        node.next = None

        return last

    def reorderList(self, head: ListNode) -> None:
        if head == None or head.next == None:
            return
        # find the middle position
        slow = head
        fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        slow = slow.next

        temp = head
        while head.next != slow:
            head = head.next
        head.next = None
        head = temp

        # reverse the linked list
        re_head = self.reverse(slow)

        # combine two linked list
        cur = head
        while cur and re_head:
            temp = cur.next
            cur.next = re_head
            re_head = re_head.next
            cur.next.next = temp
            cur = cur.next.next

File: Go/LinkedList/143-重排链表.go
package LinkedList

// 找到中间节点，反转后半部分，然后间隔插入前半部分
func reorderList(head *ListNode)  {
	// 头部添加一个额外节点，快慢指针找到中间节点
	head = &ListNode{
		Next: head,
	}
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	// 反转后半部分
	var prev *ListNode
	mid, tmp := slow.Next, slow
	slow.Next = nil
	for mid != nil {
		tmp = mid.Next
		mid.Next = prev
		prev = mid
		mid = tmp
	}
	// 间隔插入前半部分
	slow = head.Next
	for slow != nil && prev != nil {
		tmp = prev.Next
		prev.Next = slow.Next
		slow.Next = prev
		slow = prev.Next
		prev = tmp
	}
	printLL(head.Next)
}

